dio/lib/src/adapters/io_adapter.dart

```diff
@@ -187,27 +187,24 @@ class IOHttpClientAdapter implements HttpClientAdapter {
     Future<void>? cancelFuture,
     Duration? connectionTimeout,
   ) {
-    HttpClient client = onHttpClientCreate?.call(HttpClient()) ?? HttpClient();
     if (cancelFuture != null) {
+      final HttpClient client =
+          onHttpClientCreate?.call(HttpClient()) ?? HttpClient();
       client.userAgent = null;
       client.idleTimeout = Duration(seconds: 0);
       cancelFuture.whenComplete(() => client.close(force: true));
       return client..connectionTimeout = connectionTimeout;
```

Comment:
Independent of this PR...
We always seem to create a new client when we use `CancelToken`, that is pretty sad.

Reply:
I can't be sure if this is a special case, so I followed the previous logic. As I mentioned earlier, this pull request did not handle the logic of this file very well, so further thinking is needed.

Reply:
I can't be sure if this is a special case, so I followed the previous logic. As I mentioned earlier, this PR did not handle the logic of this file very well, so further thinking is needed.

Reply:
I created https://github.com/cfug/dio/issues/1811 as a follow-up.